e2e/src/tests.rs

```diff
@@ -0,0 +1,288 @@
+use std::{
+    collections::HashMap,
+    io::{BufRead, BufReader, Read},
+    path::Path,
+};
+
+use k8s_openapi::api::{batch::v1::Job, core::v1::Pod};
+use kube::{api::ListParams, Api, Client, Config};
+use tokio::time::{sleep, Duration};
+
+use crate::utils::*;
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    #[tokio::test]
+    // starts the node(express.js) api server, sends four different requests, validates data,
+    // stops the server and validates if the agent job and pod are deleted
+    async fn test_complete_node_api() {
+        let mut config = Config::infer().await.unwrap();
+        config.accept_invalid_certs = true;
+        let client = Client::try_from(config).unwrap();
+
+        let service_url = get_service_url(&client, "default").await.unwrap();
+        let pod_name = get_nginx_pod_name(&client, "default").await.unwrap();
+        let command = vec!["node", "node-e2e/app.js"];
+        let env = HashMap::from([("MIRRORD_AGENT_TTL", "0")]); // for adding more environment variables
+        let mut server = start_server(&pod_name, command, env);
+        sleep(Duration::from_secs(5)).await; // sleep to make sure the server is up
+        match server.try_wait() {
+            Ok(Some(status)) => {
+                if !status.success() {
+                    let mut stderr_reader = BufReader::new(server.stderr.take().unwrap());
+                    let mut err_stream = String::new();
+                    stderr_reader.read_to_string(&mut err_stream).unwrap();
+                    panic!("{}", err_stream);
+                } else {
+                    panic!("server exited with status code: {}", status);
+                }
+            }
+            Ok(None) => {} // server is running
+            Err(e) => panic!("{:?}", e),
```

Comment:
I'm not sure why you check that server is running here, we shouldn't care AFAIK

Reply:
for example if the server errors out for some reason on start, it will write to it's stderr and on line 44 we are reading the stdout and it will block forever, we dont exactly know what could happen so we need to check if all went i.e. server is running or it errored out. its just a guess that the server will be up and running in 5 secs (worst case)

Reply:
you should read the stdout when the process finishes, there's no value in reading it in real time.
flow should be:
1. start process
2. send requests, do stuff
3. kill/verify process is done
4. validate stderr/stdout of process